base/applications/mspaint/mouse.cpp

```diff
@@ -277,293 +508,374 @@ endPaintingL(HDC hdc, LONG x, LONG y, COLORREF fg, COLORREF bg)
-                    Poly(hdc, pointStack, pointSP, fg, bg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle(), FALSE, FALSE);
-                }
-            }
-            if (pointSP == 255)
-                pointSP--;
-            break;
-        case TOOL_ELLIPSE:
-            imageModel.ResetToPrevious();
-            if (GetAsyncKeyState(VK_SHIFT) < 0)
-                regularize(start.x, start.y, x, y);
-            Ellp(hdc, start.x, start.y, x, y, fg, bg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle());
-            break;
-        case TOOL_RRECT:
-            imageModel.ResetToPrevious();
-            if (GetAsyncKeyState(VK_SHIFT) < 0)
-                regularize(start.x, start.y, x, y);
-            RRect(hdc, start.x, start.y, x, y, fg, bg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle());
-            break;
-        case TOOL_FILL:
-        case TOOL_COLOR:
-        case TOOL_ZOOM:
-        case TOOL_BRUSH:
-        case TOOL_AIRBRUSH:
-            break;
+        }
+        ToolBase::OnButtonUp(bLeftButton, x, y);
     }
-}
 
-void
-startPaintingR(HDC hdc, LONG x, LONG y, COLORREF fg, COLORREF bg)
+    void OnCancelDraw()
+    {
+        imageModel.ResetToPrevious();
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
+
+// TOOL_LINE
+struct LineTool : ToolBase
 {
-    start.x = x;
-    start.y = y;
-    last.x = x;
-    last.y = y;
-    switch (toolsModel.GetActiveTool())
-    {
-        case TOOL_FREESEL:
-        case TOOL_TEXT:
-        case TOOL_LINE:
-        case TOOL_RECT:
-        case TOOL_ELLIPSE:
-        case TOOL_RRECT:
-            imageModel.CopyPrevious();
-            break;
-        case TOOL_RUBBER:
-            imageModel.CopyPrevious();
-            Replace(hdc, x, y, x, y, fg, bg, toolsModel.GetRubberRadius());
-            break;
-        case TOOL_FILL:
-            imageModel.CopyPrevious();
-            Fill(hdc, x, y, bg);
-            break;
-        case TOOL_PEN:
-            imageModel.CopyPrevious();
-            SetPixel(hdc, x, y, bg);
-            break;
-        case TOOL_BRUSH:
-            imageModel.CopyPrevious();
-            Brush(hdc, x, y, x, y, bg, toolsModel.GetBrushStyle());
-            break;
-        case TOOL_AIRBRUSH:
+    LineTool() : ToolBase(TOOL_LINE)
+    {
+    }
+
+    void draw(BOOL bLeftButton, LONG x, LONG y)
+    {
+        if (bLeftButton)
+            Line(m_hdc, start.x, start.y, x, y, m_fg, toolsModel.GetLineWidth());
+        else
+            Line(m_hdc, start.x, start.y, x, y, m_bg, toolsModel.GetLineWidth());
+    }
+
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnButtonDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+    }
+
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        imageModel.ResetToPrevious();
+        if (GetAsyncKeyState(VK_SHIFT) < 0)
+            roundTo8Directions(start.x, start.y, x, y);
```

Comment:
Seems this could be done within the `draw` function...

Reply:
OK.